nixos/modules/services/networking/prosody.nix

```diff
@@ -1,887 +1,675 @@
-        VirtualHost "${v.domain}"
-          enabled = ${boolToString v.enabled};
-          ${ optionalString (v.ssl != null) (createSSLOptsStr v.ssl) }
-          ${ v.extraConfig }
-        '') cfg.virtualHosts) }
-    '';
+            If you know what you're doing, you can disable this warning by
+            setting config.services.prosody.xmppComplianceSuite to false.
+          '';
+        };
+      in
+      [
+        (mkAssertion "muc" ''
+          You need to setup at least a MUC domain to comply with
+          XEP-0423
+        '')
+
+        (mkAssertion "http_upload" ''
+          You need to setup the http_upload module through
+          config.services.prosody.components to comply with
+          XEP-0423.
+        '')
+      ];
 
-    users.users.prosody = mkIf (cfg.user == "prosody") {
-      uid = config.ids.uids.prosody;
-      description = "Prosody user";
-      inherit (cfg) group;
-      home = cfg.dataDir;
+    warnings =
+      optionals (cfg.modules != { }) [ "The option `services.prosody.modules' has been and split into two separate options: `services.prosody.settings.modules_enabled' and `services.prosody.settings.modules_enabled'." ] ++
+      mapAttrsToList (k: v: "The option `services.prosody.virtualHosts.${k}.enabled' has been renamed to `services.prosody.virtualHosts.${k}.settings.enabled'.") (filterAttrs (_: v: v.enabled != null) cfg.virtualHosts) ++
+      mapAttrsToList (k: v: "The option `services.prosody.virtualHosts.${k}.ssl' has been renamed to `services.prosody.virtualHosts.${k}.settings.ssl'.") (filterAttrs (_: v: v.ssl != null) cfg.virtualHosts)
+    ;
+
+    services.prosody.settings = {
+      log = mkDefault { warn = "*syslog"; error = "*syslog"; };
+      data_path = mkForce cfg.dataDir;
+      network_backend = "event";
+      pidfile = "/run/prosody/prosody.pid";
+
+      authentication = mkDefault "internal_hashed";
+      reload_modules = mkIf (acmeHosts != []) [ "tls" ];
+
+      modules_enabled = [
+        # required for compliance with https://compliance.conversations.im/about/
+        "dialback"
+        "disco"
+        "roster"
+        "saslauth"
+        "tls"
+
+        # not essential, but recommended
+        "blocklist"
+        "bookmarks"
+        "carbons"
+        "cloud_notify"
+        "csi"
+        "pep"
+        "private"
+        "vcard_legacy"
+
+        # nice to have
+        "mam"
+        "ping"
+        "register"
+        "smacks"
+        "time"
+        "uptime"
+        "version"
+
+        # admin interfaces
+        "admin_adhoc"
+        "http_files"
+        "proxy65"
+      ] ++ cfg.package.communityModules
+        ++ mapAttrsToList (k: _: k) (filterAttrs (_: v: v == true) cfg.modules);
+
+      modules_disabled = mapAttrsToList (k: _: k) (filterAttrs (_: v: v == false) cfg.modules);
+
+      disco_items =
+        mapAttrsToList (k: v: [ k "${k} HTTP upload endpoint" ]) (filterAttrs (k: v: v.module == "http_upload") cfg.components) ++
```

Comment:
`${k} HTTP upload endpoint` used to be `HTTP upload endpoint` -- this is probably fine as it looks like a comment to me, but didn't check.

More importantly, we now have a disco_items in base of config and one in each virtualhosts, but the content isn't repeated. My config didn't have one in the virtualhost so if the vhost inherited base config that changed:
```
-disco_items = {
-disco_items = {
-{ "tmp.xxx", "HTTP upload endpoint"};
-{ "rooms.xxx", "rooms.xxx MUC endpoint"};
-};
+disco_items = { { "tmp.xxx", "tmp.xxx HTTP upload endpoint" }, { "rooms.xxx", "rooms.xxx MUC endpoint" } }

VirtualHost "xxx"
+disco_items = {  }
```

Similarly, the vhost now has empty `modules_disabled = {}` and `modules_enabled  = {}` which might or might not matter

Reply:
> this is probably fine as it looks like a comment to me, but didn't check.

Documentation states it is purely a description so I'm confident we're fine with this.

> More importantly, we now have a disco_items in base of config and one in each virtualhosts, but the content isn't repeated. My config didn't have one in the virtualhost so if the vhost inherited base config that changed:

I'm glad we have different interpretations of how prosody is passin parsing this! I assumed so confidently that inherited lists are additive  instead of not (which implies overwriting) that I never checked! I'll go back and find out if my assumption is correct and document it if so. I'm really glad you pointed this out! Thanks.

Reply:
this isn't as easy to check as i hoped... i hopped into a `prosody` shell (via `sudo -u prosody prosodyctl shell`) and called `module:list()` - it looks to me that modules are additive as i suspected so i will consider this resolved unless we specifically have a reason to not... agree?

Reply:
I hadn't checked either and just pointed out any change I could see by diffing the old and new configs -- since you've confirmed it's additive this looks good to me! Thanks!